Групповой проект версия 1.

MoriYandex opened the pull request.

muha-zhi committed:
feat: Добавить функциональность 'Рекомендации'

muha-zhi committed:
Исправил ошибки стиля

muha-zhi committed:
Checkstyle

UlianaKomissarova committed:
feat: добавление отзывов

UlianaKomissarova committed:
fix: исправление багов

m3Rzy committed:
Добавлен функционал ленты событий на добавление/удаление друзей, появление лайка/дизлайка. Убраны лишние классы -InMemory, немного заменен код и закомментированы тесты

m3Rzy committed:
Убран лишний импорт

m3Rzy committed:
Добавлено удаление Фильма с помощью filmId

m3Rzy committed:
Добавлено удаление пользователя с помощью userId

UlianaKomissarova committed:
feat: добавление общих фильмов

UlianaKomissarova committed:
fix: удаление лишнего импорта

UlianaKomissarova committed:
fix: исправление работы теста для базы

UlianaKomissarova committed:
fix: удаление ненужных импортов

m3Rzy committed:
Добавлен функционал отзывов

m3Rzy committed:
Добавлена лента новостей на все методы

UlianaKomissarova committed:
fix: получение полезности одним запросом

muha-zhi committed:
fix: Исправить ошибки add-director

muha-zhi committed:
fix: Исправил ошибку с тестом

muha-zhi committed:
Исправил ошибку с валидацией

muha-zhi committed:
Исправил ошибку стиля

muha-zhi committed:
Добавил функциональность add-director

UlianaKomissarova committed:
feat: вывод популярных фильмов с фильтрацией по жанру и году

UlianaKomissarova committed:
fix: правка бага селект-запроса

muha-zhi committed:
Добавил поиск фильмов по названию и имени режиссера

MoriYandex committed:
Merge pull request #9 from MoriYandex/add-reviews

feat: добавление отзывов

MoriYandex committed:
Merge pull request #15 from MoriYandex/add-remove-endpoint

1.1 Удаление фильмов и пользователей. 2 SP

Commit:
Merge branch 'add-recommendations' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

Commit:
Merge branch 'add-common-films' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/test/java/ru/yandex/practicum/filmorate/tests/FilmControllerTests.java

Commit:
Merge branch 'add-director' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/tests/FilmControllerTests.java

Commit:
Merge branch 'add-most-populars' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

Commit:
Merge branch 'add-search' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/tests/FilmControllerTests.java

Commit:
Merge branch 'add-feed' into develop

# Conflicts:
#	pom.xml
#	src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/review/DbReviewStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/DbUserStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java
#	src/main/resources/schema.sql
#	src/test/java/ru/yandex/practicum/filmorate/tests/FilmControllerTests.java
#	src/test/java/ru/yandex/practicum/filmorate/tests/UserControllerTests.java

Commit:
Групповой проект версия 1

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,63 @@
+    private final ReviewService reviewService;
+
+    @PostMapping
+    public Review addReview(@Valid @RequestBody Review review) {
```
Объект называется ReviewController, методы принимают/возвращают объекты типа Review, а что если убрать упоминание Review из названий методов, кажется что станет покороче без потери читаемости :)

MoriYandex replied:
В контроллерах и Storage убрал названия сущностей из методов. В сервисах решил оставить - там может быть значительно больше разных методов с разнообразными названиями.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,49 @@
+    private final DirectorStorage directorStorage;
+
+    public Director createDirector(Director director) {
+        if (director.getName().isBlank()) {
```
Если вдруг name == null, то мы получим ошибку, тут бы на null еще проверку :)

MoriYandex replied:
Заменил на StringUtils.isBlank()

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -43,6 +45,33 @@ public List<Film> getAllFilms() {
+        if (directorStorage.getDirectorById(id) == null) {
+            throw new NotFoundException(String.format("Режиссер с id %s не найден", id));
+        }
+        List<Film> allFilms = filmStorage.getAllFilms();
```
Мы тут выгребаем все фильмы из нашей базы. В общем случае их может быть очень много и у нас просто памяти не хватит :)
Давай сделаем методы на уровне хранилища

MoriYandex replied:
Выбор реализован на уровне хранилища.
Отбор лайков и жанров по ИД фильма перенесён в БД.
Сортировка оставлена на уровне Java, так как мне хотелось оставить алгоритм, согласно которому мы берём массив лайков из БД для всех отобранных фильмов, а не возвращаем из базы результат JOIN'а по лайкам - в таком результате может быть ну очень большой объём данных, и мы потеряем немало времени на маппинг.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -43,6 +45,33 @@ public List<Film> getAllFilms() {
+
+            case likes:
+                return allFilms.stream()
+                        .filter(p1 -> !p1.getDirectors().isEmpty() && p1.getDirectors().get(0).getId() == id)
```
.filter(p1 -> !p1.getDirectors().isEmpty() && p1.getDirectors().get(0).getId() == id)

Вот, тут, выборка данных - это обязанность хранилища, лучше переместить в метод хранилища :)

MoriYandex replied:
Переместил.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -64,17 +98,25 @@ public Film addLike(Integer id, Integer userId) {
+    }
+
+    public List<Film> getCommonFilms(Integer userId, Integer friendId) {
+        log.info(String.format("FilmService: Вывод общих фильмов пользователей %d и %d", userId, friendId));
```
`log.info(String.format("FilmService: Вывод общих фильмов пользователей %d и %d", userId, friendId));`

Это можно сделать вот так

`log.info("Вывод общих фильмов пользователей {} и {}", userId, friendId);`

FilmService: можно не упоминать, имя класса и так будет в логе.

Далее, в логгере лучше использовать плейсхолдеры. Это еще и более оптимально так как такие выражения не будут вычисляться вовсе в случае если не включен соответствующий уровень логирования. В случае же с конкатенацией выражения будут выполняться всегда :)

MoriYandex replied:
Логи поправил.
В качестве улучшения стоит отдельно подумать над единой системой логирования - что логировать в сервисах, а что в хранилищах, например. Но это всё-таки считаю уроком на будущее (следует заранее договориться о такой системе всей командой) - а то делать стройно и красиво можно очень долго, а дедлайн никуда не денется.

EugeneLenkevich replied:
> Логи поправил. В качестве улучшения стоит отдельно подумать над единой системой логирования - что логировать в сервисах, а что в хранилищах, например. Но это всё-таки считаю уроком на будущее (следует заранее договориться о такой системе всей командой) - а то делать стройно и красиво можно очень долго, а дедлайн никуда не денется.

Ну тут главное не перемудрить. Главное чтобы при разборе ошибки с прода по логам модно было понять что там происходило :)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -96,4 +121,13 @@ public void validateUser(User user) {
     }
+
+    private boolean isExist(int id) {
+        for (User user : userStorage.getAllUsers()) {
```
Опять таки, не стоит вытаскивать все записи из таблицы, у нас есть СУБД, она гораздо эффективней может проверять наличие записей

https://stackoverflow.com/questions/7471625/fastest-check-if-row-exists-in-postgresql

MoriYandex replied:
Логично. Поправил.

MoriYandex replied:
Пришлось проводить рефакторинг целых разделов, но я его провёл.
Очень хотелось сделать одним запросом рекомендации. Но почему-то не получалось никак. Уже даже в Postgres проверял запрос. Может, h2 неправильно обрабатывает трёхэтажные конструкции With?

EugeneLenkevich replied:
> Пришлось проводить рефакторинг целых разделов, но я его провёл. Очень хотелось сделать одним запросом рекомендации. Но почему-то не получалось никак. Уже даже в Postgres проверял запрос. Может, h2 неправильно обрабатывает трёхэтажные конструкции With?

В таких случаях надо постараться написать запрос на JPQL. Диалекты H2 и Postgres отличаются. Как вариант - запускать постгрес в докере в тестах, но это слегка вне рамок курса наверно :)

MoriYandex replied:
Хотелось бы с вами списаться по этому поводу уже после сдачи проекта...

MoriYandex replied:
Опытным путём установлено, что H2 не воспринимает конструкции более чем с одним подзапросом WITH. Сделал с одним - всё заработало и стало красиво.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java:
```diff
@@ -17,5 +18,14 @@ public interface FilmStorage {
     Film deleteLike(Integer id, Integer userId);
 
-    List<Film> getMostPopular(Integer count);
+    //List<Film> getMostPopular(Integer count);
```
Закомментированный код менее читаем и понятен для других разработчиков, что может привести к ошибкам и проблемам при компиляции или выполнении программы. Поэтому рекомендуется всегда удалять закомментированный код из своего кода программы.

Ну и, скорее всего, все вокруг поменяется и он станет не актуальным, а для истории есть git :)

MoriYandex replied:
Убрал. Комментарий был создан при слиянии, забыл убрать потом.

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -43,6 +45,33 @@ public List<Film> getAllFilms() {
     }
 
+    public List<Film> getFilmsByDirId(Integer id, FilmSortBy by) {
+        if (directorStorage.getDirectorById(id) == null) {
```
Тут, кстати, можно сделать вот такой метод, который вернет boolean, так проще :)

`directorStorage.existsById(id)`

MoriYandex replied:
Метод такой нашёлся. Только переименовать пришлось.

EugeneLenkevich reviewed (changes requested):
Привет, Мориниллэ! Спасибо за PR.

Отличная работа, спасибо! :)

1. Логику выборки и сортировки данных лучше по максимуму выполнять на стороне СУБД, она это делает гораздо эффективней
2. Хотелось бы немножко поправить логи, комментарий оставил
3. Еще можно подумать над именованием методов классов, вдруг получится записать покороче. 

Но, в целом, все супер :)

Commit:
Групповой проект версия 2 - rename

Commit:
Групповой проект версия 2 - logs

Commit:
Групповой проект версия 2 - storages

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)

Commit:
Групповой проект версия 2 - Рекомендации.
Опытным путём установлено, что БД H2 не поддерживает множественные подзапросы WITH - а вот с одиночным всё ок.
